Evidently integration (standard steps and example)

alex-zenml opened the pull request:
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
- imported `rich` for some pretty printing to the terminal (ahead of next week's import)
- fixed bug where `Output` objects could only include multiple items
- added standard interface steps for the six core `Profile` outputs from `evidently`
- added an example to show the functionality in action
- added integration tests for the `drift_detection` example

AlexejPenner commented on examples/drift_detection/README.md:
````diff
@@ -0,0 +1,88 @@
+### Pre-requisites
+In order to run this example, you need to install and initialize ZenML:
+
+```shell
+# install CLI
+pip install zenml
+
+# pull example
+zenml example pull drift_detection
+```
````
I think it would make sense to expose the user to the more familiar way of calling python scripts instead of advertising zenml example run. This way the user can make changes to the script and rerun to see how the script behaves.

```suggestion
```shell
# install CLI
pip install zenml

# install ZenML integrations
zenml integration install evidently
zenml integration install sklearn

# pull example
zenml example pull drift_detection
cd zenml_examples/drift_detection

# initialize
zenml init
```
```

AlexejPenner commented on examples/drift_detection/README.md:
````diff
@@ -0,0 +1,88 @@
+Now we're ready. Execute:
+
+```shell
+zenml example run drift_detection
````
```suggestion
python run.py
```

AlexejPenner commented on src/zenml/steps/utils.py:
```diff
@@ -411,6 +411,9 @@ def Do(
         if return_type is not None:
             if isinstance(return_type, Output):
                 # Resolve named (and multi-) outputs.
+                # breakpoint()
```
Was this comment intended to stay in here

AlexejPenner commented on src/zenml/steps/utils.py:
```diff
@@ -411,6 +411,9 @@ def Do(
             if isinstance(return_type, Output):
                 # Resolve named (and multi-) outputs.
+                # breakpoint()
+                if sum(1 for dummy_var in return_type.items()) == 1:
```
Is there a reason len(return_type.items()) == 1 does not work here?

alex-zenml replied:
You can't call `len()` on a generator.

AlexejPenner replied:
Totally makes sense, wasn't aware its a generator :)

AlexejPenner reviewed (commented):
examples/drift_detection/__init__.py is missing a Copyright statement, i couldn't comment the file directly

alex-zenml commented:
@AlexejPenner addressed / fixed everything you mentioned.

schustmi commented on src/zenml/steps/utils.py:
```diff
@@ -411,6 +411,8 @@ def Do(
         if return_type is not None:
             if isinstance(return_type, Output):
                 # Resolve named (and multi-) outputs.
+                if sum(1 for dummy_var in return_type.items()) == 1:
```
Three super small things:
- Usually `_` is the name indicating that the variable is not used
- Iterating over a dict directly (instead of `.keys()`, `.items()` or `.values()` is the same as iterating over the keys, so in this case I think removing the `.items()` looks cleaner
- I don't remember why but I don't see any reason why the `output.items()` should be a generator, so I'd suggest changing that to a dict and using `len()` here as Alexej suggested
-

jwwwb replied:
what about just `len(list(return_type))` ? that seems clearest to me

alex-zenml replied:
@schustmi Got it for the first two. 

For the third comment, when I run our `quickstart.py` with the `len` implementation, I get the following error:

```
Traceback (most recent call last):
  File "examples/quickstart/quickstart.py", line 75, in <module>
    pipeline.run()
  File "/Users/strickvl/coding/zenml/repos/zenml/src/zenml/pipelines/base_pipeline.py", line 324, in run
    ret = self.stack.orchestrator.run(self, run_name=run_name)
  File "/Users/strickvl/coding/zenml/repos/zenml/src/zenml/orchestrators/local/local_orchestrator.py", line 54, in run
    runner.run(tfx_pipeline, run_name)
  File "/Users/strickvl/coding/zenml/repos/zenml/src/zenml/orchestrators/local/local_dag_runner.py", line 112, in run
    execute_step(component_launcher)
  File "/Users/strickvl/coding/zenml/repos/zenml/src/zenml/orchestrators/utils.py", line 67, in execute_step
    execution_info = tfx_launcher.launch()
  File "/Users/strickvl/.pyenv/versions/zenml-dev-3.8.6/lib/python3.8/site-packages/tfx/orchestration/portable/launcher.py", line 545, in launch
    executor_output = self._run_executor(execution_info)
  File "/Users/strickvl/.pyenv/versions/zenml-dev-3.8.6/lib/python3.8/site-packages/tfx/orchestration/portable/launcher.py", line 420, in _run_executor
    executor_output = self._executor_operator.run_executor(execution_info)
  File "/Users/strickvl/.pyenv/versions/zenml-dev-3.8.6/lib/python3.8/site-packages/tfx/orchestration/portable/python_executor_operator.py", line 133, in run_executor
    return run_with_executor(execution_info, executor)
  File "/Users/strickvl/.pyenv/versions/zenml-dev-3.8.6/lib/python3.8/site-packages/tfx/orchestration/portable/python_executor_operator.py", line 58, in run_with_executor
    result = executor.Do(execution_info.input_dict, output_dict,
  File "/Users/strickvl/coding/zenml/repos/zenml/src/zenml/steps/utils.py", line 415, in Do
    if len(return_type.items()) == 1:
TypeError: object of type 'generator' has no len()
```

That error is fixed with my implementation as above.

schustmi replied:
Ah my bad `return_type` isn't actually a dict, it's a separate object and `.items()` is something that we implemented so we still need to use `return_type.items()`. For my third part I was referring to actually changing the code inside the `StepOutput` class, so that `StepOutput.items()` doesn't return a generator anymore but a dictionary.
I also liked James' suggestion of implementing `__len__` and potentially even some for of indexing on the `StepOutput` class

AlexejPenner commented:
> @AlexejPenner addressed / fixed everything you mentioned.

Thanks @strickvl , overall the rest of your implementation looks great, and runs smoothly on my machine :)

stefannica reviewed (changes requested):
Some suggestions that I think will make this integration easier to maintain and more flexible in the long run

alex-zenml commented on src/zenml/artifacts/data_drift_artifact.py:
```diff
@@ -0,0 +1,21 @@
+from zenml.artifacts.base_artifact import BaseArtifact
+
+
+class DataDriftArtifact(BaseArtifact):
```
You're totally right. I had originally anticipated using this, but then decided it would have just added unnecessary complexity. When / if we have more monitoring integrations, we might consider standardising them via a class like you suggest, but I think I / we don't know enough what would go into that right now.

I have removed this class & references to it.

alex-zenml commented on src/zenml/integrations/evidently/steps/evidently_probabilistic_model_performance.py:
```diff
@@ -0,0 +1,70 @@
+        Returns:
+            a dict containing the results of the model performance
+        """
+        probabilistic_model_performance_profile = Profile(
+            sections=[ProbClassificationPerformanceProfileSection()]
+        )
+        probabilistic_model_performance_profile.calculate(
+            reference_dataset,
+            comparison_dataset,
+            column_mapping=config.column_mapping,
+        )
```
Fully agree with this idea, though less confident about how to implement. I'll take a stab at it along the lines you suggest.

alex-zenml commented:
@stefannica Implemented your suggestions in these latest commits.

schustmi commented on src/zenml/materializers/built_in_materializer.py:
```diff
@@ -32,6 +32,7 @@ class BuiltInMaterializer(BaseMaterializer):
         str,
         bytes,
         dict,
+        Dict,
```
If this solved the issue for dictionary annotations, does it make sense to include `typing.List` and `typing.Tuple` here as well?

alex-zenml replied:
Actually it didn't. This should maybe be removed. I am just using `dict` still in my implementation.

schustmi replied:
And maybe add a TODO so we can take a look at that, I think that's quite an important issue as `Dict`, `List` are the "correct" way to annotate your function in python <= 3.9 I believe

stefannica commented on src/zenml/integrations/evidently/steps/evidently_profile.py:
```diff
@@ -0,0 +1,112 @@
+                f"Invalid profile section: {self.profile_section}"
+                "\n\n"
+                "Valid and supported options are: \n"
+                "- datadrift, \n"
+                "- categoricaltargetdrift, \n"
+                "- numericaltargetdrift, \n"
+                "- classificationmodelperformance, \n"
+                "- regressionmodelperformance, \n"
+                "- probabilisticmodelperformance \n"
```
I like how descriptive this error message it. You should use `profile_mapper.keys()` to populate this list and save yourself the trouble of having to keep them in sync.

alex-zenml replied:
I had initially tried to do that, but couldn't figure out how to add those individual items as arguments for the `KeyError` error raising. I'll try again. I agree that it would be better.

stefannica commented on src/zenml/integrations/evidently/steps/evidently_profile.py:
```diff
@@ -0,0 +1,112 @@
+            )
+
+    column_mapping: Optional[ColumnMapping]
+    profile_section: str
```
Entirely optional: model this as a `Sequence[str]` instead, which nicely maps 1-to-1 with what the evidently Profile accepts as constructor argument. This gives the user even more flexibility.

alex-zenml replied:
Yes great idea. Have added this in.

stefannica reviewed (approved):
Thanks Alex, this looks great. I added a couple of minor suggestions, if you have the time.

stefannica commented on src/zenml/steps/utils.py:
```diff
@@ -411,6 +411,8 @@ def Do(
         if return_type is not None:
             if isinstance(return_type, Output):
                 # Resolve named (and multi-) outputs.
+                if len(list(return_type)) == 1:
```
Something I forgot to mention in my last review: the last time I tested your branch, this change raised an exception for me, because the `Output` type is not iterable. Maybe I'm missing some other changes that landed in main in the meantime, but you might want to check on your side.

stefannica replied:
`list(return_type.items())` instead of `list(return_type)` fixed it for me btw

alex-zenml commented:
@stefannica all suggestions implemented now. Barring any failing tests, I think we should be good to go now. 🤞

htahir1 commented on pyproject.toml:
```diff
@@ -60,6 +60,7 @@ pydantic = "<=1.8.2"
 analytics-python = "^1.4.0"
 distro = "^1.6.0"
 tabulate = "^0.8.9"
+rich = "^11.0.0"
```
Do we already want this in even before we're using it? :-) I would keep it out for now

alex-zenml replied:
I'm actually using it in the `drift_detection` example.

htahir1 commented on src/zenml/integrations/evidently/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    for ZenML."""
+
+    NAME = EVIDENTLY
+    REQUIREMENTS = ["evidently"]
```
Lets pin it to the latest version and above

htahir1 commented on src/zenml/integrations/evidently/steps/evidently_profile.py:
```diff
@@ -0,0 +1,107 @@
+        except KeyError:
+            nl = "\n"
+            raise ValueError(
+                f'Invalid profile section: {self.profile_section} \n\nValid and supported options are: {nl}- {f"{nl}- ".join(list(profile_mapper.keys()))}'
```
This needs formatting to 80 characters doesnt it?

htahir1 commented on src/zenml/materializers/built_in_materializer.py:
```diff
@@ -26,6 +26,9 @@
 class BuiltInMaterializer(BaseMaterializer):
     """Read/Write JSON files."""
 
+    # TODO [LOW]: consider adding typing.Dict and typing.List
```
Are they really the `correct` way of doing it? I thought this was a thing that was required for backwards compatibility reasons

jwwwb replied:
I think the more abstract`typing.Mapping` is reccomended for function parameters (generic key value map is accepted) but `typing.Dict` as a more specific annotation for the return type

htahir1 commented on examples/drift_detection/README.md:
```diff
@@ -0,0 +1,111 @@
+set and compares it against another comparison dataset. These are both input in
+the form of a `pandas` dataframe, though CSV inputs are also possible.
+
+ZenML implements this functionality in the form of several standardized steps. You select which of the profile sections you want to use in your step by passing a string into the `EvidentlyProfileConfig`. Possible options supported by Evidently are:
```
Maybe add a link break to conform to the rest of the files line length?

htahir1 commented on examples/drift_detection/run.py:
```diff
@@ -0,0 +1,95 @@
+logger = get_logger(__name__)
+
+
+@step()
```
No need for brackets here

htahir1 commented on examples/drift_detection/run.py:
```diff
@@ -0,0 +1,95 @@
+    drift_detector(reference_dataset=full_data, comparison_dataset=partial_data)
+
+
+pipeline = drift_detection_pipeline(
```
Lets nest everything from here on in a `if __name__ == '__main__'` as a standard good practice

htahir1 commented on tests/integration/test_examples.py:
```diff
@@ -13,6 +13,7 @@
 #  permissions and limitations under the License.
 import shutil
 from pathlib import Path
+from typing import Dict
```
Didn't we decide to go away from this `Dict` stuff to `dict`? I remember @schustmi saying that we should just use `dict` but there seems to have been a change for some reason. Maybe I missed it though

schustmi replied:
See my answer above, here with the official documentation (https://docs.python.org/3.9/library/typing.html#typing.Dict)

htahir1 reviewed (changes requested):
I have not run this yet but wanted to give an initial quick review. Will loop back with another one on this and fileio in the next hour

alex-zenml commented:
@htahir1 addressed all your comments/suggestions in the latest commits. (aside from the `dict` one). (@stefannica note that you'll want to work off these latest changes, though they shouldn't affect your work, I think)

alex-zenml commented on src/zenml/steps/step_output.py:
```diff
@@ -26,3 +26,7 @@ def __init__(self, **kwargs: Type[Any]):
         """Yields a tuple of type (output_name, output_type)."""
         yield from self.outputs.__annotations__.items()
+
+    def __len__(self) -> int:
```
I thought it was being used when we call `len()` on an `Output` step? Maybe I'm misunderstanding something.

schustmi replied:
Correct! However I think there's a small issue here, the `__len__` function in this case returns the length of the first output element right? Shouldn't it return the amount of specified outputs instead?

alex-zenml replied:
I've deleted it. Trying to fix caused some other unexpected errors and not enough time to methodically think this through.

htahir1 reviewed (approved):
LGTM!
